File: LeetCode_Algorithm/270(Closest Binary Search Tree Value)/270.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
'''
LeetCode 270 : Closest Binary Search Tree Value
트리 노드에서 타겟에 가장 가까운 값을 찾아라

# sol1 53ms 18% 16.1MB 74%
bfs로 순회하여 답 도출 

# sol2


'''
from typing import Optional
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        # sol1
        ans = 0
        abs_val = float("inf")
        from collections import deque
        queue = deque([root])
        while queue:
            node = queue.popleft()
            if abs(node.val - target) < abs_val:
                abs_val = abs(node.val - target)
                ans = node.val
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        return ans

        
                

File: LeetCode_Algorithm/270(Closest Binary Search Tree Value)/270.js
var closestValue = function(root, target) {
  // sol1 80ms 35% 45.9MB 17%
  let ans = -1; 
  let absVal = Infinity
  let queue = [root]
  while (queue.length) {
    let node = queue.shift()
    if (absVal > Math.abs(node.val - target)) {
      absVal = Math.abs(node.val - target)
      ans = node.val
    }
    if (node.left)
      queue.push(node.left)
    if (node.right)
      queue.push(node.right)
  }
  return ans
};